File: Common/PolyPaint.Common/Collaboration/DrawViewModel.cs
using Newtonsoft.Json;
using Newtonsoft.Json.Serialization;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows;

namespace PolyPaint.Common.Collaboration
{
    [Serializable]
    public class DrawViewModel
    {
        public string Guid { get; set; }
        public string Owner { get; set; }
        public ItemTypeEnum ItemType { get; set; }
        public List<PolyPaintStylusPoint> StylusPoints { get; set; }
        public PolyPaintColor FillColor { get; set; }
        public PolyPaintColor BorderColor { get; set; }
        public double BorderThickness { get; set; }
        public string BorderStyle { get; set; }
        public string ShapeTitle { get; set; }
        public List<string> Methods { get; set; }
        public List<string> Properties { get; set; }
        public string SourceTitle { get; set; }
        public string DestinationTitle { get; set; }
        public string ChannelId { get; set; } = "general";
        public string OutilSelectionne { get; set; }
        public PolyPaintStylusPoint LastElbowPosition { get; set; }
        public byte[] ImageBytes { get; set; }
        public double Rotation { get; set; }
        // Guid, Side
        public List<List<string>> InConnections;
        public List<List<string>> OutConnections;

        [JsonIgnore]
        public List<PolyPaintStylusPoint> RotatedStylusPoints
        {
            get { return StylusPoints.ToList().Select(x => RotatePoint(new PolyPaintStylusPoint(x.X, x.Y), Center, -Rotation)).ToList(); }
        }

        [JsonIgnore]
        public PolyPaintStylusPoint Center
        {
            get { return new PolyPaintStylusPoint((StylusPoints[1].X + StylusPoints[0].X) / 2, (StylusPoints[1].Y + StylusPoints[0].Y) / 2); }
        }

        public bool IsConnection()
        {
            var connections = new List<ItemTypeEnum>()
            {
                ItemTypeEnum.Agregation,
                ItemTypeEnum.Composition,
                ItemTypeEnum.Line,
                ItemTypeEnum.UnidirectionalAssociation,
                ItemTypeEnum.Inheritance,
                ItemTypeEnum.BidirectionalAssociation,
            };
            return connections.Contains(ItemType);
        }

        [JsonIgnore]
        private readonly JsonSerializerSettings SerializerSettings = new JsonSerializerSettings
        {
            ContractResolver = new DefaultContractResolver(),
            NullValueHandling = NullValueHandling.Ignore
        };

        public override string ToString() => JsonConvert.SerializeObject(this, SerializerSettings);

        public static PolyPaintStylusPoint RotatePoint(PolyPaintStylusPoint p1, PolyPaintStylusPoint p2, double angle)
        {
            double radians = ConvertToRadians(angle);
            double sin = Math.Sin(radians);
            double cos = Math.Cos(radians);

            // Translate point back to origin
            p1.X -= p2.X;
            p1.Y -= p2.Y;

            // Rotate point
            double xnew = p1.X * cos - p1.Y * sin;
            double ynew = p1.X * sin + p1.Y * cos;

            // Translate point back
            PolyPaintStylusPoint newPoint = new PolyPaintStylusPoint(xnew + p2.X, ynew + p2.Y);
            return newPoint;
        }

        public static double ConvertToRadians(double angle)
        {
            return (Math.PI / 180) * angle;
        }

        public static double ConvertToDegrees(double angle)
        {
            return angle * (180 / Math.PI);
        }
    }
}

File: Frontend/ios/ios/Canvas/Collaboration/Common/DrawViewModel.swift
//
//  DrawViewModel.swift
//  ios
//
//  Created by William Sevigny on 2019-03-06.
//  Copyright © 2019 LOG3000 equipe 12. All rights reserved.
//
import UIKit
struct DrawViewModel: Codable {
    var Guid: String?
    var Owner: String?
    var ItemType: ItemTypeEnum?
    var StylusPoints: [PolyPaintStylusPoint]?
    var FillColor: PolyPaintColor?
    var BorderColor: PolyPaintColor?
    var BorderThickness: Double?
    var BorderStyle: String?
    var ShapeTitle: String?
    var Methods: [String]?
    var Properties: [String]?
    var SourceTitle: String?
    var DestinationTitle: String?
    var ChannelId: String?
    var OutilSelectionne: String?
    var LastElbowPosition: PolyPaintStylusPoint?
    var ImageBytes: [UInt8]?
    var Rotation: Double?

    init() {}
}

enum ItemTypeEnum: Int, Codable {
    case Activity
    case Artefact
    case Phase
    case Comment
    case Role
    case UmlClass
    case Text
    case Agregation
    case BidirectionalAssociation
    case Composition
    case Inheritance
    case UniderectionalAssoication
    case Image
    case Line
    
    var description: String {
        switch self {
        case .Activity:
            return "Activity"
        case .Artefact:
            return "Artefact"
        case .Phase:
            return "Phase"
        case .Comment:
            return "Comment"
        case .Role:
            return "Role"
        case .UmlClass:
            return "Class"
        case .Text:
            return "Text"
        case .Agregation:
            return "Connection"
        case .BidirectionalAssociation:
            return "Connection"
        case .Composition:
            return "Connection"
        case .Inheritance:
            return "Connection"
        case .UniderectionalAssoication:
            return "Connection"
        case .Image:
            return "Image"
        case .Line:
            return "Connection"
        }
    }
}

struct PolyPaintStylusPoint: Codable {
    var X: Double
    var Y: Double
    var PressureFactor: Float
    
    init(X: Double, Y: Double, PressureFactor: Float) {
        self.X = X
        self.Y = Y
        self.PressureFactor = PressureFactor
    }
    
    init(point: CGPoint) {
        self.X = Double(point.x)
        self.Y = Double(point.y)
        self.PressureFactor = 1
    }
    
    func getCGPoint() -> CGPoint {
        return CGPoint(x: self.X, y: self.Y)
    }
}

struct PolyPaintColor: Codable {
    var A: Int
    var R: Int
    var G: Int
    var B: Int
    
    init(A: Int, R: Int, G: Int, B:Int) {
        self.A = A
        self.R = R
        self.G = G
        self.B = B
    }
    
    init(color: UIColor)  {
        var red: CGFloat = 0
        var green: CGFloat = 0
        var blue: CGFloat = 0
        var alpha: CGFloat = 0
        color.getRed(&red, green: &green, blue: &blue, alpha: &alpha)
        
        self.A = alpha <= 1 ? Int(alpha * 255): Int(alpha)
        self.R = red <= 1 ? Int(red * 255): Int(red)
        self.G = green <= 1 ? Int(green * 255): Int(green)
        self.B = blue <= 1 ? Int(blue * 255): Int(blue)
    }
    
    func getUIColor() -> UIColor {
        print(A, R, G, B)
        return UIColor(red: CGFloat(self.R), green: CGFloat(self.G), blue: CGFloat(self.B), alpha: CGFloat(self.A))
    }
}
